File: app/src/main/java/com/juawapps/newsspread/data/api/adapters/UnwrapResponseConverter.java
package com.juawapps.newsspread.data.api.adapters;

import android.support.annotation.NonNull;

import com.juawapps.newsspread.data.api.ApiStatus;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Converter;

/**
 * Unwraps the object response from the api to the actual object.
 */

class UnwrapResponseConverter<T>
        implements Converter<ResponseBody, T> {

    private final Converter<ResponseBody, WrappedApiResponse<T>> converter;

    UnwrapResponseConverter(Converter<ResponseBody,
            WrappedApiResponse<T>> converter) {
        this.converter = converter;
    }

    @Override
    public T convert(@NonNull ResponseBody value) throws IOException {
        WrappedApiResponse<T> response = converter.convert(value);

        if (response.getStatus().equalsIgnoreCase(ApiStatus.STATUS_OK)) {
            return response.getContent();
        }

        throw new ApiError("");
    }

}

File: app/src/main/java/com/juawapps/newsspread/data/api/adapters/UnwrapResponseConverter.kt
package com.juawapps.newsspread.data.api.adapters

import com.juawapps.newsspread.data.api.ApiStatus

import java.io.IOException

import okhttp3.ResponseBody
import retrofit2.Converter

/**
 * Unwraps the object response from the api to the actual object.
 */

internal class UnwrapResponseConverter(private val converter: Converter<ResponseBody?, *>?) : Converter<ResponseBody, Any> {

    @Throws(IOException::class)
    override fun convert(value: ResponseBody): Any? {
        val response = converter?.convert(value)

        if (response is WrappedApiResponse<*> && response.status.equals(ApiStatus.STATUS_OK, ignoreCase = true)) {
            return response.content
        }

        throw ApiError("")
    }

}

